### ISSUE TITLE
Expose state even when not building with inline

### ISSUE THREAD

**Issue by vlovich** [Cyan4973/xxHash#359]
There's lot of reason to want to allocate the struct on the stack (e.g. inside of a semi hotloop) without needing to allocate. I understand the concerns around it but maybe the restriction can be relaxed when a XXH_NAMESPACE is defined? Seems unlikely you'd end up with linkage issues, no?

**Comment by easyaspi314** [comment]
We already support that, just do this:
```c
#define XXH_STATIC_LINKING_ONLY
#include "xxhash.h"
```

**Comment by vlovich** [comment]
Perhaps I didn't clarify. I explicitly don't want to inline the methods themselves. Those are better left as defined within xxhash.c . I just want to be able to use the regular version that doesn't bloat the binary with stack-allocated state.

**Comment by vlovich** [comment]
Oh NVM. I think I misread the docs.

**Comment by easyaspi314** [comment]
I'll explain.

`XXH_STATIC_LINKING_ONLY` doesn't inline everything. It unlocks declarations of internals that have not been stabilized. `XXH_INLINE_ALL` implies this.

This prevents issues with dynamic linking:

You allocate an `XXH64_state_t` on the stack and compile against xxHash 0.7.3. You link against `libxxhash.so`.

The current `sizeof(XXH64_state_t)` (should) be 88 bytes, so the compiler will allocate 88 bytes on the stack:

```asm
hash:
    push    rbp
    mov     rbp, rsp
    sub     rsp, 88 # XXH64_state_t state;
    mov     rdi, rsp # &state
    movabs  rsi, 0x1234567890abcdef # seed
    call    XXH64_resetState@PLT
    mov     rdi, rsp # &state 
    lea     rsi, [rip+.Lhello_world] # "Hello, world"
    mov     edx, 12 # strlen("Hello, world")
    call    XXH64_update@PLT
    mov     rdi, rsp # &state
    call    XXH64_digest@PLT # rax = XXH64_digest(&state)
    mov     rsp, rbp
    pop     rbp
    # return rax
    ret
```
This works fine on xxHash 0.7.3.

However, for some reason, in v0.8.0, we add two 64-bit values to `XXH64_state_t`, bringing it to 104 bytes.

The 88 bytes is still hardcoded into the binary, but the library expects 104 bytes. It goes out of bounds, overwrites the return address, and crashes the program or does something worse.

`XXH_INLINE_ALL` does not allow `XXH_NAMESPACE` because there is no reason to. `XXH_NAMESPACE` is to prevent public symbol collisions, which doesn't happen with inlines.

**Comment by vlovich** [comment]
I understand the ABI compatability concerns but I figured most people defining a namespace would also be ensuring only 1 version of the library gets linked. Thanks for clarifying the distinctions of the different macros. There's a lot of them & I overlooked that `XXH_STATIC_LINKING_ONLY ` doesn't statically define the implementations in the header (for which both `XXH_INLINE_ALL` and `XXH_PRIVATE_API` do & also imply `XXH_STATIC_LINKING_ONLY`).
